src/charm_state.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""State of the Charm."""
+import itertools
+import typing
+
+# pydantic is causing this no-name-in-module problem
+from pydantic import (  # pylint: disable=no-name-in-module,import-error
+    BaseModel,
+    Extra,
+    Field,
+    ValidationError,
+    validator,
+)
+
+from exceptions import CharmConfigInvalidError
+
+if typing.TYPE_CHECKING:
+    from charm import SynapseCharm
+
+
+KNOWN_CHARM_CONFIG = (
+    "server_name",
+    "report_stats",
+)
+
+
+class SynapseConfig(BaseModel):  # pylint: disable=too-few-public-methods
+    """Represent Synapse builtin configuration values.
+
+    Attrs:
+        server_name: server_name config.
+        report_stats: report_stats config.
+    """
+
+    server_name: str | None = Field(None, min_length=0)
+    report_stats: str | None = Field(None)
+
+    class Config:  # pylint: disable=too-few-public-methods
+        """Config class.
+
+        Attrs:
+            extra: extra configuration.
+        """
+
+        extra = Extra.allow
+
+    @validator("server_name")
+    # no-self-argument disabled due to cls usage as expected by validator
+    def check_server_name_empty(cls, value: str) -> str:  # pylint: disable=no-self-argument
```

Comment:
Isn't the `server_name: str | None = Field(None, min_length=0)` already validate that server name can't be empty?

Reply:
Yes, I changed the min_length, thanks